pkg/istio/client.go

```diff
@@ -118,60 +103,58 @@ func (c *Client) createVirtualService(ctx context.Context, watcher *v1alpha1.Wat
 }
 
-func (c *Client) lookupGateway(ctx context.Context, watcher *v1alpha1.Watcher) (*istioclientapi.Gateway, error) {
-	gName, gSel := mergeSelectorCfg(c.config.Gateway, watcher.Spec.Gateway)
-
-	if gName == "" && gSel == nil {
-		c.eventRecorder.Event(watcher, "Warning", "WatcherGatewayNotConfigured",
-			"Watcher: Gateway for the VirtualService not configured")
-		return nil, errNoGatewayConfigured
-	}
-
-	ls, err := metav1.LabelSelectorAsSelector(gSel)
+func appendHosts(gateways []*istioclientapi.Gateway, virtualSvc *istioclientapi.VirtualService) error {
+	hosts, err := getHosts(gateways)
 	if err != nil {
-		return nil, fmt.Errorf("error converting label selector: %w", err)
+		return err
 	}
-	lss := ls.String()
-
-	// Gateway namespacedName takes precedence as it is more specific than label selector lookup
-	if gName != "" {
-		namespace, name := splitOnSlash(gName)
-		gateway, err := c.NetworkingV1beta1().
-			Gateways(namespace).
-			Get(ctx, name, metav1.GetOptions{})
-		if err != nil {
-			return nil, fmt.Errorf("error getting configured istio gateway: %w", err)
-		}
+	virtualSvc.Spec.Hosts = append(virtualSvc.Spec.Hosts, hosts...)
+	return nil
+}
 
-		return gateway, nil
+func getHosts(gateways []*istioclientapi.Gateway) ([]string, error) {
+	uniqueHostsMap := make(map[string]bool)
+	for i := range gateways {
+		gateway := gateways[i]
+		servers := gateway.Spec.Servers
+		if len(servers) == 0 || len(servers[0].Hosts) == 0 {
+			return nil, fmt.Errorf("for gateway %s: %w",
+				client.ObjectKeyFromObject(gateway).String(),
+				ErrCantFindGatewayServersHost)
+		}
+		// The first Hosts entry must be the listener host, use map to filter out duplicate hosts
+		uniqueHostsMap[servers[0].Hosts[0]] = true
+	}
+	hosts := make([]string, 0)
+	for host := range uniqueHostsMap {
+		hosts = append(hosts, host)
 	}
+	return hosts, nil
+}
 
-	// fallback to label selector
-	lo := metav1.ListOptions{
-		LabelSelector: lss,
+func (c *Client) LookupGateways(ctx context.Context, watcher *v1alpha1.Watcher) ([]*istioclientapi.Gateway, error) {
+	selector, err := metav1.LabelSelectorAsSelector(&watcher.Spec.Gateway.LabelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("error converting label selector: %w", err)
 	}
+	labelSelector := selector.String()
 	gateways, err := c.NetworkingV1beta1().
 		Gateways(metav1.NamespaceAll).
-		List(ctx, lo)
+		List(ctx, metav1.ListOptions{
+			LabelSelector: labelSelector,
+		})
 	if err != nil {
-		return nil, fmt.Errorf("error looking up Istio gateway with the label selector %q: %w", lss, err)
+		return nil, fmt.Errorf("error looking up Istio gateway with the label selector %q: %w",
+			labelSelector, err)
 	}
 
 	if len(gateways.Items) == 0 {
 		c.eventRecorder.Event(watcher, "Warning", "WatcherGatewayNotFound",
 			"Watcher: Gateway for the VirtualService not found")
-		return nil, fmt.Errorf("%w. Label selector: %q", errCantFindMatchingGateway, lss)
-	}
-
-	if len(gateways.Items) > 1 {
```

Comment:
Why is this removed? According to the https://github.com/kyma-project/lifecycle-manager/issues/234, we should only support one gateway. Is the contract changed?

Reply:
Yes, here the original AC is not a valid concern IMO, I exchanged this idea with @jakobmoellersap already. And if you check the istio virtual service definition https://istio.io/latest/docs/reference/config/networking/virtual-service/, `Spec.Gateways` is a list, it's not necessary to allow only one, and the decision in the AC to use the first find one is also a bit premature:
First, I don't think it can always fetch the same gateway if have multiple selected, the list is not sorted.
Second, if any of the gateway configured wrong and this gateway get selected, by only emitting warning but not stop the process could be even more hard to do troubleshooting.

I think I mentioned about this in daily, for clarification, any specific concern to only allow one gateway? @khlifi411

Reply:
I have no concerns from my side about having the vs pointing to many gateways.